sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala

```diff
@@ -546,14 +548,23 @@ case class FlatMapGroupsInRWithArrowExec(
 
   override protected def doExecute(): RDD[InternalRow] = {
     child.execute().mapPartitionsInternal { iter =>
-      val grouped = GroupedIterator(iter, groupingAttributes, child.output)
+      val grouped = GroupedIterator(iter, groupingAttributes, child.output).filter(_._2.hasNext)
       val getKey = ObjectOperator.deserializeRowToObject(keyDeserializer, groupingAttributes)
-      val runner = new ArrowRRunner(func, packageNames, broadcastVars, inputSchema,
-        SQLConf.get.sessionLocalTimeZone, RRunnerModes.DATAFRAME_GAPPLY)
 
-      val groupedByRKey = grouped.map { case (key, rowIter) =>
-        val newKey = rowToRBytes(getKey(key).asInstanceOf[Row])
-        (newKey, rowIter)
+      // Iterating over keys is relatively cheap.
+      val keys: Iterator[Array[Byte]] =
+        grouped.map { case (key, rowIter) => rowToRBytes(getKey(key).asInstanceOf[Row]) }
+      val groupedByRKey: Iterator[Iterator[InternalRow]] =
+        grouped.map { case (key, rowIter) => rowIter }
+
+      val runner = new ArrowRRunner(func, packageNames, broadcastVars, inputSchema,
+        SQLConf.get.sessionLocalTimeZone, RRunnerModes.DATAFRAME_GAPPLY) {
+        protected override def bufferedWrite(
+            dataOut: DataOutputStream)(writeFunc: ByteArrayOutputStream => Unit): Unit = {
+          super.bufferedWrite(dataOut)(writeFunc)
+          // Don't forget we're sending keys additionally.
+          keys.foreach(dataOut.write)
```

Comment:
This logic is changed too. I tried to use `ArrowRRunner` to only send Arrow batches, and send keys separately here.

Reply:
so we expect `keys` is small?

Reply:
It's a bit hard for me to visualize... isnt each row associated with a key, which could be different? if I understand before it's a pairing for `(key, row)` or `(key, rowIter)` - how does it handle this now?

Reply:
Yea, keys are expected to be small. I am also going to try to explain from scratch to make sure we're synced.

Input iterator:

```
[a, 1]
[b, 2]
[b, 3]
```

Output iterator:

```
([a], Iterator([a, 1])
([b], Iterator([b, 2], [b, 3])
```

Here, I am doing:

```
next() call => ([a], Iterator([a, 1])
  keeps the key 'a'
  converts Iterator([a, 1]) to Arrow batch
  buffers the Arrow batch
next() call => ([b], Iterator([b, 2], [b, 3])
  keeps the key 'a'
  converts Iterator([b, 2], [b, 3]) to Arrow batch
  buffers the Arrow batch
...

Write buffered all Arrow batches
Write kept all the keys.
```

I tried to only leave Arrow conversion logics in `ArrowRRunner` (which is shared for both `dapply` and `gapply`), and keeping/writing keys logic is here separate.

(BTW, I tried to avoid those refactorings cus I know it's difficult to review such PRs).

Reply:
does `[a]` `[b]` mean in your illustration an array of 1?

Reply:
Oh, right. The types look confusing.

Input iterator:

`Iterator[InternalRow]`

Output Iterator

`Iterator[(InternalRow, Iterator[InternalRow])]`

```bash
next() call => ([a], Iterator([a, 1])             # (InternalRow, Iterator[InternalRow])
  1. converts the key 'a' into R-readable bytes   # 'a' becomes InternalRow -> Array[Byte] (see [1])
  2. keeps the converted key in `keys`            # 'keys` is mutable.ArrayBuffer[Array[Byte] (see [1])
  3. converts Iterator([a, 1]) to Arrow batch     # Iterator[InternalRow] -> Array[Byte]
  4. buffers the Arrow batch                      # buffer: Array[Byte] + Arrow batch: Array[Byte]
...
Write buffered all Arrow batches                  # See [2]
Write all the keys.                               # See [3]
```

[1] https://github.com/apache/spark/pull/23977/files#diff-a1240a7ba20d0e027d23a690a770bf44R557
[2] https://github.com/apache/spark/pull/23977/files#diff-a0b6a11cc2e2299455c795fe3c96b823R71
[3] https://github.com/apache/spark/pull/23977/files#diff-a1240a7ba20d0e027d23a690a770bf44R567

Reply:
ok thanks for the explanation